### PR TITLE
Add Kubernetes StatefulSet metricset

### PR DESCRIPTION
**Pull request by fassisrosa**
Added support for Kubernetes StatefulSet.

### CODE REVIEW

File: metricbeat/module/kubernetes/state_statefulset/data.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package state_statefulset
```
[the remark this answers is not in the document]
Comment (fassisrosa) [reply metricbeat/module/kubernetes/state_statefulset/data.go]: Following same pattern as other classes.

**Comment by fassisrosa** [comment]
Added unit testing for StatefulSet support. Made StatefulSet be relative to namespace (i.e. you can have two statefulsets with same name belonging to different namespaces).

**Comment by fassisrosa** [comment]
I signed CLA already.

File: metricbeat/module/kubernetes/state_statefulset/data.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/module/kubernetes/state_statefulset/data.go]: yes, please ignore the bot on this.

**Comment by ruflin** [comment]
@fassisrosa Thanks for you contribution. Haven't look at the code yet but it for sure will need an entry in the CHANGELOG :-)

**Comment by fassisrosa** [comment]
Added entry to CHANGELOG.

File: metricbeat/module/kubernetes/state_statefulset/data.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+func eventMapping(families []*dto.MetricFamily) ([]common.MapStr, error) {
+	eventsMap := map[string]common.MapStr{}
+	for _, family := range families {
+		for _, metric := range family.GetMetric() {
+			statefulset := util.GetLabel(metric, "statefulset")
+			if statefulset == "" {
+				continue
+			}
+			namespace := util.GetLabel(metric, "namespace")
+			statefulset_key := namespace + "::" + statefulset
```
[the remark this answers is not in the document]
Comment (exekias) [reply metricbeat/module/kubernetes/state_statefulset/data.go]: I think the bot comment is valid here :innocent: 

Now I see this, it seems the rest of `state_*` may miss the namespacing you did here, I'm glad you found this!

File: metricbeat/module/kubernetes/state_statefulset/state_statefulset.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+type MetricSet struct {
+	mb.BaseMetricSet
+	prometheus *helper.Prometheus
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The kubernetes state_statefulset metricset is beta")
```
Comment (exekias) [inline metricbeat/module/kubernetes/state_statefulset/state_statefulset.go]: You can remove this message, kubernetes module has been promoted to GA

**Comment by exekias** [comment]
jenkins, test it please

File: metricbeat/module/kubernetes/state_statefulset/state_statefulset_test.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+						testValue(t, event, k, v)
+					}
+					delete(testCases, eventKey)
+				}
+			}
+		}
+	}
+
+	if len(testCases) > 0 {
+		t.Error("Test reference events not found: %v", testCases)
```
Comment (exekias) [inline metricbeat/module/kubernetes/state_statefulset/state_statefulset_test.go]: This should be Errorf

File: metricbeat/module/kubernetes/state_statefulset/data.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+			case "kube_statefulset_status_replicas":
+				event.Put("replicas.observed", metric.GetGauge().GetValue())
+			default:
+				// Ignore unknown metric
+				continue
+			}
+		}
+	}
+
+	var events []common.MapStr
```
Comment (walktall) [inline metricbeat/module/kubernetes/state_statefulset/data.go]: Maybe you could pre allocate events by `events := make([]common.MapStr, 0, len(eventsMap))`

File: metricbeat/module/kubernetes/state_statefulset/state_statefulset.go (hunk 3 above)
Comment (fassisrosa) [reply metricbeat/module/kubernetes/state_statefulset/state_statefulset.go]: Done.

Comment (fassisrosa) [reply metricbeat/module/kubernetes/state_statefulset/state_statefulset.go]: Fixed bot comment.

File: metricbeat/module/kubernetes/state_statefulset/state_statefulset_test.go (hunk 4 above)
Comment (fassisrosa) [reply metricbeat/module/kubernetes/state_statefulset/state_statefulset_test.go]: Done.

File: metricbeat/module/kubernetes/state_statefulset/data.go (hunk 5 above)
Comment (fassisrosa) [reply metricbeat/module/kubernetes/state_statefulset/data.go]: Done.

File: metricbeat/module/kubernetes/state_statefulset/data.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fassisrosa) [reply metricbeat/module/kubernetes/state_statefulset/data.go]: I was going to mention it after this pull request. Worth fixing others no doubt. :)

**Comment by fassisrosa** [comment]
Any idea what this Travis CI build fail is all about? Looks like a code convention violation but  have no idea about the test’s expectations...

**Comment by walktall** [comment]
@fassisrosa You should obey the the go import style. First section with std pkgs, second is community pkgs such as github.com, third section is project pkg in this case is pkg under beats project.

**Comment by fassisrosa** [comment]
Understood, I think... but then how did state_replicaset_test.go pass checks? Tried to follow same convention. 

Just to confirm, issue is in state_statefulset_test.go imports should be in correct order:

 +.    "net/http"
 +	"net/http/httptest"
 +	"os"
 +	"testing"
 +
 +	"github.com/stretchr/testify/assert"
 +
 +	"github.com/elastic/beats/libbeat/common"
 +	mbtest "github.com/elastic/beats/metricbeat/mb/testing"
 
Is this the correct fix?

Thanks for your patience, this is my first commit to beats, trying to understand the rules for submissions :)

**Comment by exekias** [comment]
That looks correct, you can use `make fmt` to automatically fix formatting

**Comment by fassisrosa** [comment]
Thanks @exekias , made changes, ran fmt, pushed.

**Comment by exekias** [comment]
ok to test

File: metricbeat/module/kubernetes/state_statefulset/state_statefulset.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	prometheus *helper.Prometheus
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	return &MetricSet{
+		BaseMetricSet: base,
+		prometheus:    helper.NewPrometheusClient(base),
```
Comment (exekias) [inline metricbeat/module/kubernetes/state_statefulset/state_statefulset.go]: This line is failing in the CI, `NewPrometheusClient` can return an error

Comment (exekias) [reply metricbeat/module/kubernetes/state_statefulset/state_statefulset.go]: To put some context, this changed very recently: https://github.com/elastic/beats/pull/6205, probably your working copy doesn't have the change

**Comment by exekias** [comment]
Thank you so much for this great addition @fassisrosa :tada:!!

**Comment by fassisrosa** [comment]
Thanks so much for the review comments. Love metricbeat, using it all the time! :)

**Comment by exekias** [comment]
I've opened https://github.com/elastic/beats/issues/6281 to fix the namespacing issue in the rest of `state_` metricsets, plan to take it next week, let me know if you want to take a stab at it :smile_cat:  https://github.com/elastic/beats/issues/6281

**Comment by fassisrosa** [comment]
Taking a stab at it... Pull request above. :)

**Pull request merged**